controllers/fluxcd/multi-cluster-controller.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2022 The KubeSphere Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fluxcd
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"kubesphere.io/devops/pkg/api/gitops/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+//+kubebuilder:rbac:groups=cluster.kubesphere.io,resources=clusters,verbs=get;list;watch
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;create;update
+
+// MultiClusterReconciler represents a controller to sync cluster to ArgoCD cluster
+type MultiClusterReconciler struct {
+	client.Client
+	log      logr.Logger
+	recorder record.EventRecorder
+}
+
+// Reconcile sync the cluster.kubesphere.io/clusters to kubeconfig secret
+// in DevopsProject Namespaces where Flux Application(HelmRelease and Kustomization) exist.
+func (r *MultiClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+	r.log.Info(fmt.Sprintf("start to reconcile member cluster kubeconfig: %s", req.String()))
+
+	cluster := createBareClusterObject()
+	if err = r.Get(ctx, req.NamespacedName, cluster); err != nil {
+		err = client.IgnoreNotFound(err)
+		return
+	}
+
+	// ignore the host cluster
+	if ignore(cluster) {
+		return
+	}
+
+	if err = r.reconcileCluster(cluster); err != nil {
+		return
+	}
+
+	return
+}
+
+func ignore(cluster *unstructured.Unstructured) bool {
+	if cluster != nil {
+		_, ok := cluster.GetLabels()["cluster-role.kubesphere.io/host"]
+		return ok
+	}
+	return true
+}
+
+func (r *MultiClusterReconciler) reconcileCluster(cluster *unstructured.Unstructured) (err error) {
+	ctx := context.Background()
+	name := cluster.GetName()
+	nsList := &v1.NamespaceList{}
+	if err = r.getFluxAppNsList(ctx, nsList); err != nil {
```

Comment:
Please reuse the context from https://github.com/kubesphere/ks-devops/pull/769/files#diff-8c80dfd76c63967c04928983c1b74b1e26e00ac2b1dcbd790030457c4f147b4cR51

Reply:
Thanks. I'll fix this.